File: src/application/cachePolicy.ts
export interface CachedValue<T> {
  readonly value: T;
  readonly storedAt: Date;
}

export interface CachePolicyOptions<T> {
  readonly load: () => Promise<T>;
  readonly readCache: () => Promise<CachedValue<T> | null>;
  readonly writeCache: (value: T) => Promise<void>;
  readonly ttlMs: number;
  readonly now?: () => Date;
  readonly networkAvailable?: () => boolean;
}

/**
 * L'unique implémentation du stale-while-revalidate du projet (ADR-010, principe
 * repris d'ADR-008). Lecture du cache → rendu immédiat → si le TTL est dépassé
 * ET que le réseau répond, rafraîchissement en tâche de fond.
 *
 * Ne jamais recopier cette logique dans un dépôt ni dans un écran : c'est
 * exactement ce que la décision d'architecture interdit. Chaque recopie est une
 * divergence future.
 *
 * ⚠️ Le TTL est celui du **cache**. Il n'a rien à voir avec la fraîcheur d'une
 * observation, qui se mesure sur `date_obs` aux seuils absolus de BR-005
 * (2 h / 24 h) — voir `domain/observation/freshness.ts`.
 */
export function withCachePolicy<T>(options: CachePolicyOptions<T>): () => Promise<T> {
  const now = options.now ?? (() => new Date());
  const networkAvailable = options.networkAvailable ?? (() => true);

  /**
   * Rafraîchissement en vol, s'il y en a un. Sans cette mémoire, N lectures
   * simultanées sur une entrée expirée déclenchent N appels réseau — un écran
   * de carte en produit autant qu'il affiche de stations. Hub'Eau n'annonce
   * aucun quota (`C-15`) : rien ne nous arrêterait, c'est donc à nous de le
   * faire, et ce composant est le seul endroit qui puisse le faire.
   */
  let enVol: Promise<void> | null = null;

  return async function read(): Promise<T> {
    const cached = await options.readCache();

    if (cached === null) {
      const fresh = await options.load();
      await options.writeCache(fresh);
      return fresh;
    }

    // La borne appartient à l'état périmé : on ne prolonge jamais un cache.
    const expired = now().getTime() - cached.storedAt.getTime() >= options.ttlMs;

    if (expired && networkAvailable() && enVol === null) {
      // Volontairement non attendu : l'affichage ne doit pas dépendre du réseau.
      // Un échec de rafraîchissement laisse la dernière valeur connue en place
      // plutôt que de vider l'écran (BR-007).
      enVol = options
        .load()
        .then((fresh) => options.writeCache(fresh))
        .catch(() => undefined)
        // Libéré dans tous les cas : un rafraîchissement en échec ne doit pas
        // bloquer définitivement les suivants.
        .finally(() => {
          enVol = null;
        });
      void enVol;
    }

    return cached.value;
  };
}

File: lib/application/cache_policy.dart
// La politique de cache vit ici, et nulle part ailleurs (cf. CLAUDE.md §
// Architecture) : jamais recopiee dans un depot ni dans un ecran. Rendu
// immediat depuis le cache, rafraichissement en tache de fond
// (stale-while-revalidate) si l'entree est perimee et le reseau disponible.
// L'affichage n'attend jamais le reseau.

import 'dart:async';

/// Une valeur mise en cache, avec l'instant de son ecriture. [storedAt] est
/// la date de RECUPERATION par le cache — sans rapport avec la fraicheur
/// metier d'une observation (BR-005, qui date l'observation elle-meme, pas
/// le cache qui la contient). La borne de peremption appartient a l'etat
/// perime : c'est [withCachePolicy] qui la lit, jamais un depot ni un ecran.
final class CachedValue<T> {
  const CachedValue({required this.value, required this.storedAt});

  /// La valeur mise en cache.
  final T value;

  /// L'instant ou [value] a ete ecrite en cache.
  final DateTime storedAt;
}

bool _toujoursDisponible() => true;

/// Construit une fonction de lecture stale-while-revalidate.
///
/// Rend immediatement la valeur en cache si [readCache] en renvoie une, sans
/// jamais attendre [load] pour l'affichage. Si cette valeur est perimee
/// (`now() - storedAt >= ttl`) et que [networkAvailable] repond `true`, lance
/// [load] en tache de fond pour rafraichir le cache via [writeCache].
///
/// Deduplique les rafraichissements en vol (`C-12`) : plusieurs lectures
/// simultanees sur une entree perimee ne declenchent qu'un seul appel a
/// [load]. Un rafraichissement en echec laisse la derniere valeur connue,
/// n'ecrit rien, et ne bloque pas les rafraichissements suivants — la mise
/// en vol est liberee dans un `finally` (BR-007).
///
/// Un cache vide (`readCache` rend `null`) attend [load], le rend et
/// l'ecrit : rien n'est encore perime, il n'y a rien de perime a afficher en
/// attendant.
///
/// Leve un [ArgumentError] si [ttl] n'est pas strictement positif — a la
/// construction de la fonction, avant tout appel.
Future<T> Function() withCachePolicy<T>({
  required Future<T> Function() load,
  required Future<CachedValue<T>?> Function() readCache,
  required Future<void> Function(T value) writeCache,
  required Duration ttl,
  DateTime Function()? now,
  bool Function()? networkAvailable,
}) {
  if (ttl <= Duration.zero) {
    throw ArgumentError.value(ttl, 'ttl', 'doit etre strictement positif');
  }

  final DateTime Function() horloge = now ?? DateTime.now;
  final bool Function() reseauDisponible =
      networkAvailable ?? _toujoursDisponible;

  Future<T>? rafraichissementEnVol;

  Future<T> rafraichir() {
    final Future<T>? existant = rafraichissementEnVol;
    if (existant != null) {
      return existant;
    }

    Future<T> lancer() async {
      try {
        final T valeur = await load();
        await writeCache(valeur);
        return valeur;
      } finally {
        rafraichissementEnVol = null;
      }
    }

    final Future<T> lance = lancer();
    rafraichissementEnVol = lance;
    return lance;
  }

  return () async {
    final CachedValue<T>? enCache = await readCache();
    if (enCache == null) {
      return rafraichir();
    }

    final bool perime = horloge().difference(enCache.storedAt) >= ttl;
    if (perime && reseauDisponible()) {
      unawaited(
        rafraichir().then<void>(
          (T _) {},
          onError: (Object _) {
            // La derniere valeur connue reste en cache (BR-007) : l'echec
            // est deja absorbe ici, il ne doit pas ressortir comme erreur
            // non geree.
          },
        ),
      );
    }

    return enCache.value;
  };
}
